framework/src/org/checkerframework/framework/type/BoundsInitializer.java

```diff
@@ -391,36 +465,76 @@ public void initializeExtendsBound(final AnnotatedWildcardType wildcard) {
+                    TypeMirror javaTypeArg =
+                            getJavaType(declaredType, typeElement.getTypeParameters(), i);
+                    final AnnotatedTypeMirror atmArg =
+                            AnnotatedTypeMirror.createType(
+                                    javaTypeArg, declaredType.atypeFactory, false);
+                    typeArgs.add(atmArg);
+                    if (atmArg.getKind() == TypeKind.WILDCARD && declaredType.wasRaw()) {
+                        ((AnnotatedWildcardType) atmArg).setUninferredTypeArgument();
                     }
                 }
-                declaredType.setTypeArguments(typeArgs);
             } else {
+                typeArgs = declaredType.typeArgs;
+            }
 
-                final List<AnnotatedTypeMirror> typeArgs =
-                        new ArrayList<>(declaredType.getTypeArguments());
-                final List<AnnotatedTypeMirror> typeArgReplacements =
-                        new ArrayList<>(typeArgs.size());
-                for (int i = 0; i < typeArgs.size(); i++) {
-                    final AnnotatedTypeMirror typeArg = typeArgs.get(i);
-                    final BoundPathNode node = addPathNode(new TypeArgNode(i));
-                    typeArgReplacements.add(replaceOrVisit(typeArg));
-                    removePathNode(node);
+            List<AnnotatedTypeMirror> typeArgReplacements = new ArrayList<>(typeArgs.size());
+            for (int i = 0; i < typeArgs.size(); i++) {
+                AnnotatedTypeMirror typeArg = typeArgs.get(i);
+                BoundPathNode node = addPathNode(new TypeArgNode(i));
+                if (typeArg.getKind() == TypeKind.WILDCARD) {
+                    ((AnnotatedWildcardType) typeArg)
+                            .setTypeVariable(typeElement.getTypeParameters().get(i));
                 }
+                typeArgReplacements.add(replaceOrVisit(typeArg));
+                removePathNode(node);
+            }
+
+            declaredType.setTypeArguments(typeArgReplacements);
+        }
 
-                declaredType.setTypeArguments(typeArgReplacements);
+        /**
+         * Store the wildcards created as type arguments to raw types. This way if the raw type is
+         * visited again, the same wildcard can be used.
+         */
+        private Map<TypeVariable, WildcardType> rawTypeWildcards = new HashMap<>();
+
+        /**
+         * Returns the underlying java type of the {@code i} type parameter of {@code type}. If
+         * {@code type} is raw, then a new wildcard is created or returned from {@code
+         * rawTypeWildcards}.
+         *
+         * @param type declared type
+         * @param parameters elements of the type parameters
+         * @param i index of the type parameter
+         * @return the underlying java type of the {@code i} type parameter of {@code type}
+         */
+        private TypeMirror getJavaType(
```

Comment:
Can you please move `getJavaType` and `getJavaTypeArg` next to each other? Their javadoc is basically the same, but this version caches the created wildcard.
How does one choose between the two? Are the wildcards computed differently? Why are both needed?
Add a note that the implementation of the two should be kept in sync?